inc/azure_macro_utils/macro_utils.h

```diff
@@ -78,73 +78,101 @@ MU_IF(X, "true", "false") => "true"
 #define MU_EAT_EMPTY_ARG(arg_count, arg) MU_IF(MU_ISEMPTY(arg),,arg) MU_IF(MU_ISEMPTY(arg),MU_EXPAND_TO_NOTHING,MU_IFCOMMALOGIC)(MU_DEC(arg_count))
 #define MU_EAT_EMPTY_ARGS(...) MU_FOR_EACH_1_COUNTED(MU_EAT_EMPTY_ARG, __VA_ARGS__)
 
+#define MU_EMPTY()
+#define MACRO_UTILS_DELAY(id) id MU_EMPTY MU_LPAREN )
+
+//
+//
+// MU_DEFINE_ENUM
+//
+//
+
+// This uber enum macro can handle enumeration-constants being constant expressions.
+// Given the following enum:
+//
+//
+// #define TEST_UBER_ENUM_VALUES \
+//     test_uber_a, \
+//     test_uber_b = 5
+// 
+// MU_DEFINE_ENUM(TEST_UBER_ENUM, TEST_UBER_ENUM_VALUES)
+//
+// The ToString expands to something like the below given an enum
+//
+// const char* TEST_UBER_ENUM_ToString(TEST_UBER_ENUM value)
+// {
+//     static int my_test_uber_a;
+//     static int my_test_uber_b = 5;
+//     static int enum_last_value = -1;
+//     enum_last_value = my_test_uber_a ? (my_test_uber_a) : (enum_last_value + 1);
+//     if (enum_last_value == value)
+//     {
+//         return "test_uber_a";
+//     }
+//     enum_last_value = (my_test_uber_b = 5) ? (my_test_uber_b = 5) : (enum_last_value + 1);
+//     if (enum_last_value == value)
+//     {
+//         return "test_uber_a";
+//     }
+//     return "NULL";
+// }
+
 #define MU_DEFINE_ENUMERATION_CONSTANT(x) x,
-/*MU_DEFINE_ENUM_WITHOUT_INVALID goes to header*/
-#define MU_DEFINE_ENUM_WITHOUT_INVALID(enumName, ...) typedef enum MU_C2(enumName, _TAG) { MU_FOR_EACH_1(MU_DEFINE_ENUMERATION_CONSTANT, __VA_ARGS__)} enumName; \
-    extern const char* MU_C2(enumName,Strings)(enumName value); \
-    extern int MU_C2(enumName, _FromString)(const char* enumAsString, enumName* destination);
 
-// this macro is a wrapper on top of MU_DEFINE_ENUM_WITHOUT_INVALID, adding an _INVALID value as the first enum value in the enum
+#define MU_INTERNAL_DECLARE_ENUM_VAR(enumValue) \
+    static int MU_C2(my_, enumValue);
+
+#define MU_INTERNAL_ASSIGN_ENUM_VALUE(enumValue) \
+    enum_last_value = (MU_C2(my_, enumValue)) ? (MU_C2(my_, enumValue)) : (enum_last_value + 1); \
+    if (enum_last_value == value) \
+    { \
+        return MU_TOSTRING(enumValue); \
```

Comment:
>MU_TOSTRING [](start = 15, length = 11)

I would argue that a different macro is needed here. The reason is that **this**macro will print blindly the "whatever is as an enum constant". Such a "test_not_uber_all_2 = test_not_uber_all_1 +1". Leaving the user wondering what the value really is... (see SF enums, they are all build like "current = previous +1," pattern). 

I believe seeing the "real" value is of value here. If only there'd be way to extract the enumeration-constant from "enumeration-constant = constant-expression" (C11 6.7.2.2 - Enumeration specifiers).

Reply:
Ideally we could format as "identifier (value)".
Up to now we were printing "identifier". Now we print "identifier = value" if the identifier is assigned a value. This is more information than before.
If we can format it nicer later that's all good, but I think this is more than enough for now.

We can always parse the string and make it fancy, but I'd rather avoid that...

---
In reply to: [328355869](https://github.com/Azure/azure-macro-utils-c/pull/30#discussion_r328355869) [](ancestors = 328355869)